pkg/ccl/streamingccl/streamproducer/event_stream.go

```diff
@@ -324,6 +332,48 @@ func (p *checkpointPacer) shouldCheckpoint(
 	return false
 }
 
+// Add a RangeFeedSSTable into current batch and return size of the bytes added.
+func (s *eventStream) addSST(sst *roachpb.RangeFeedSSTable, batch *streampb.StreamEvent_Batch) (int, error) {
+	for _, sp := range s.spec.Spans {
+		if sp.Contains(sst.Span) {
+			batch.Ssts = append(batch.Ssts, *sst)
+			return sst.Size(), nil
+		}
+	}
+
+	// If span of the sst exceeds boundaries of the watched spans,
+	// we trim the sst data to avoid sending unnecessary data.
+	iter, err := storage.NewMemSSTIterator(sst.Data, true)
+	if err != nil {
+		return 0, err
+	}
+	defer iter.Close()
+	size := 0
+	for _, sp := range s.spec.Spans {
+		iter.SeekGE(storage.MVCCKey{Key: sp.Key})
+		endKey := storage.MVCCKey{Key: sp.EndKey}
+		for ; ; iter.Next() {
+			if ok, err := iter.Valid(); err != nil {
+				return size, err
+			} else if !ok {
+				break
+			} else if !iter.UnsafeKey().Less(endKey) {
+				break
+			}
+			keyValue := roachpb.KeyValue{
+				Key: iter.UnsafeKey().Key,
```

Comment:
I'm pretty sure it is not safe to call UnsafeKey.  I'm also surprised that none of the tests failed.  UnsafeKey returns "unsafe" buffer -- buffer that may be reset on the next call to iter.Next -- yet, you're storing these values inside batch.